### ISSUE TITLE
migrate from Travis to GitHub actions

### ISSUE THREAD

**Issue by bbolker** [glmmTMB/glmmTMB#655]
https://deanattali.com/blog/migrating-travis-to-github/

**Comment by mebrooks** [comment]
It took me a while to read this, but I'm convinced. How should we proceed? Sorry I've been MIA for a while, but I think I should be able to work on glmmTMB more in the near future. I plan to take off Dec 24-26, but will be working on and off during other days. I'll try to catch up.

**Comment by bbolker** [comment]
I took a first crack at this in the [github_actions branch](https://github.com/glmmTMB/glmmTMB/tree/github_actions) (not working yet)

**Comment by bbolker** [comment]
Mostly enabled by [this PR](https://github.com/glmmTMB/glmmTMB/pull/657), but I'm going to change the title of this issue and leave it open for future enhancements:

- disable checking for a given push/commit with a `[ci skip]` specification in the commit message (this is fairly easy to implement, [see here](https://github.com/marketplace/actions/ci-skip-action); the main thing I don't like about this particular solution is that the result shows as "failed" rather than "skipped"
- *allow* checking across all platforms with a different commit message specification (e.g. `[ci all_platforms]`)? Right now I have disabled checking except on R-release.  (This should be fairly straightforward too ...)

**Comment by bbolker** [comment]
[This answer](https://stackoverflow.com/a/65545568/190277) gives a reasonable brute-force solution to checking all platforms only under certain conditions. Don't know if there will be any hassles involved in getting a full LaTeX installation on MacOS/Windows ? Maybe skip vignette building there if necessary?

**Comment by strengejacke** [comment]
FWIW, I have copied the yaml-file from a repo from Hadley or someone related, which seems to be pretty stable: https://github.com/easystats/insight/blob/master/.github/workflows/R-check.yaml

Not sure if it helps you in your particular case here.

**Comment by bbolker** [comment]
Thanks, but I'm not sure we need it.  Checking on 9 different platforms (MacOS/Windows/Linux X oldrel/release/devel) seems like overkill for routine checking.

**Comment by strengejacke** [comment]
True, though it depends on the context. Sometimes, certain results slightly differ on different systems (like 0.037 vs. 0.041), and you don't want to wait for CRAN to complain, but fix the issues before submissions. ;-)

And sometimes, functions / packages are available for R release and devel, but not (or no longer) for old-release. Maybe it makes sense to pick at least an OS that you don't have for testing, and test and old-rel and devel, but that really depends on what is really needed and personal preferences. In particular, as those tests sometimes fail for no reason and you cannot reproduce the errors from GitHub CI locally, nor does it fail on CRAN.

**Comment by mmaechler** [comment]
We see Ben is an environmentally minded Canadian .. (as opposed to US-American -ike lazy who only cares for convenience ;-)
I strongly agree on having this __CI on opt-in only__ scheme and will propose it also to my @emacs-ess (https://github.com/emacs-ess) collaborators.

**Comment by bbolker** [comment]
Sorry to disappoint you @mmaechler, but: I'm actually suggesting the following protocol:

- default: run tests (only) on Linux, R-release
- `[skip ci]`: skip tests
- `[full ci]`: run tests on 9 platforms {(oldrel, release, devel) X (Linux, Windows, MacOS)}

Running the basic tests takes "only" about 10 minutes of compute time, I think that's acceptable
There are variety of methods of achieving this goal (see [aforementioned Stack Overflow Q&A for more details](https://stackoverflow.com/questions/65543489/github-actions-conditional-config):

- run tests and/or full tests on separate branches (advantage: no 'clutter' in commit messages; disadvantage: maintain more complicated branch structure)
- set up multiple YAML files in the workflows directory, with appropriate conditionals in each one (this is what I'm currently trying); advantage = simplicity, disadvantage = repeated code (!DRY) and two entries in the actions log for every push (either 'regular' or 'full' is skipped, but both show up in the log)
- set up a single YAML file with multiple jobs, appropriate conditionals
- set up a pre-job to define the 'job matrix' (see SO threads); this is the 'best' (from a DRY point of view) but also the most complex

**Comment by bbolker** [comment]
Next issue, getting a working TeX/TeXlive installation on all platforms. Installing texlive + packages is the easiest/most comprehensive thing, but I'm not sure I know the right incantations for making that work on all three platforms (easy on Linux). Installing tinytex + tlmgr_install any necessary packages (see [here](https://github.com/bbolker/McMasterPandemic/blob/7ca0c438367adef4aa3430316b6c6511518210fd/.github/workflows/R-CMD-check.yaml)) might work better cross-platform?

**Comment by bbolker** [comment]
ugh, still struggling with this.

- working my way through getting all required LaTeX packages installed
- Ubuntu runs are failing not able to find a LaTeX installation??? (something about adding /bin to the path??
